xmake/rules/vala/xmake.lua

```diff
@@ -58,67 +58,82 @@ rule("vala.build")
+                    table.insert(argv, "--pkg")
+                    table.insert(argv, path(package))
+                end
             end
-        end
-        if target:is_binary() then
-            for _, dep in ipairs(target:orderdeps()) do
-                if dep:is_shared() or dep:is_static() then
-                    local vapifile = dep:data("vala.vapifile")
-                    if vapifile then
-                        table.join2(argv, path(vapifile))
+            if target:is_binary() then
+                for _, dep in ipairs(target:orderdeps()) do
+                    if dep:is_shared() or dep:is_static() then
+                        local vapifile = dep:data("vala.vapifile")
+                        if vapifile then
+                            table.join2(argv, path(vapifile))
+                        end
                     end
                 end
+            else
+                local vapifile = target:data("vala.vapifile")
+                if vapifile then
+                    table.insert(argv, path(vapifile, function (p) return "--vapi=" .. p end))
+                end
+                local headerfile = target:data("vala.headerfile")
+                if headerfile then
+                    table.insert(argv, "-H")
+                    table.insert(argv, path(headerfile))
+                end
             end
-        else
-            local vapifile = target:data("vala.vapifile")
-            if vapifile then
-                table.insert(argv, path(vapifile, function (p) return "--vapi=" .. p end))
+            local vapidir = target:data("vala.vapidir")
+            if vapidir then
+                table.insert(argv, path(vapidir, function (p) return "--vapidir=" .. p end))
             end
-            local headerfile = target:data("vala.headerfile")
-            if headerfile then
-                table.insert(argv, "-H")
-                table.insert(argv, path(headerfile))
+            local valaflags = target:data("vala.flags")
+            if valaflags then
+                table.join2(argv, valaflags)
             end
+            table.insert(argv, path(sourcefile_vala))
         end
-        local vapidir = target:data("vala.vapidir")
-        if vapidir then
-            table.insert(argv, path(vapidir, function (p) return "--vapidir=" .. p end))
-        end
-        local valaflags = target:data("vala.flags")
-        if valaflags then
-            table.join2(argv, valaflags)
-        end
-        table.insert(argv, path(sourcefile_vala))
+
+        batchcmds:show_progress(opt.progress, "${color.build.object}compiling.vala")
         batchcmds:vrunv(valac.program, argv)
-        batchcmds:compile(sourcefile_c, objectfile)
+
+        for _, sourcefile_c in ipairs(sourcefiles_c) do
+            batchcmds:show_progress(opt.progress, "${color.build.object}compiling.c %s", sourcefile_c)
+            batchcmds:compile(sourcefile_c, target:objectfile(sourcefile_c))
```

Comment:
Using before_buildcmd_files will not compile them in parallel.

Reply:
The problem is that you can't compile vala code separately

Imagine simple example:
```vala
// main.vala
int main() {
    myfunc(); // valac will throw error on this line
}
```

```vala
// b.vala
void myfunc() { stdout.print("impossible"); }
```

Vala compiler has no way of knowing if there's a `myfunc` defined anywhere if compiling files separately, that's why I'm making this PR

Reply:
But here you are also compiling them one by one instead of passing a list of source files to vala all at once, so how does vala know t their symbols?

Can you provide a full test project as attachment? I will look at it.

Reply:
I was and am passing all vala files to compiler by adding them to argv

But, I refactored it and vala still has to go through entire project, but generated C code should be incrementally build now (and now Vala can properly directly import C symbols)

For project look at `tests/projects/vala/includec/src/main.vala` from my latest commit